File: app/src/main/java/io/github/ovso/righttoknow/ui/main/news/NewsAdapter.java
package io.github.ovso.righttoknow.ui.main.news;

import android.text.Html;
import android.text.SpannableString;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;
import butterknife.BindView;
import io.github.ovso.righttoknow.R;
import io.github.ovso.righttoknow.framework.utils.Utility;
import io.github.ovso.righttoknow.framework.adapter.BaseAdapterView;
import io.github.ovso.righttoknow.framework.adapter.BaseRecyclerAdapter;
import io.github.ovso.righttoknow.ui.main.news.model.News;
import java.util.ArrayList;
import java.util.List;

/**
 * Created by jaeho on 2017. 9. 1
 */

public class NewsAdapter extends BaseRecyclerAdapter
    implements NewsAdapterDataModel, BaseAdapterView {
  private List<News> items = new ArrayList<>();

  @Override protected BaseViewHolder createViewHolder(View view, int viewType) {
    return new NewsViewHolder(view);
  }

  @Override public int getLayoutRes(int viewType) {
    return R.layout.fragment_news_item;
  }

  @Override public void onBindViewHolder(BaseViewHolder viewHolder, int position) {
    if (viewHolder instanceof NewsViewHolder) {
      NewsViewHolder holder = (NewsViewHolder) viewHolder;
      final News news = items.get(position);
      String title = news.getTitle();
      SpannableString spannableString = new SpannableString(Html.fromHtml(title));
      holder.titleTextview.setText(spannableString);
      holder.countTextview.setText(String.valueOf(position+1));
      String date = Utility.convertDate(news.getPubDate(), "yy-MM-dd");
      holder.dateTextView.setText(date);
      holder.itemView.setOnClickListener(new View.OnClickListener() {
        @Override public void onClick(View view) {
          if (onRecyclerItemClickListener != null) {
            onRecyclerItemClickListener.onItemClick(news);
          }
        }
      });
      holder.imageView.setOnClickListener(
          new View.OnClickListener() {
            @Override public void onClick(View view) {
              onRecyclerItemClickListener.onSimpleNewsItemClick(news);
            }
          });
    }
  }

  @Override public int getItemCount() {
    return getSize();
  }

  @Override public void add(News item) {
    items.add(item);
  }

  @Override public void addAll(List<News> items) {
    this.items.addAll(items);
  }

  @Override public News remove(int position) {
    return items.remove(position);
  }

  @Override public News getItem(int position) {
    return items.get(position);
  }

  @Override public void add(int index, News item) {
    items.add(index, item);
  }

  @Override public int getSize() {
    return items.size();
  }

  @Override public void clear() {
    items.clear();
  }

  @Override public void refresh() {
    notifyDataSetChanged();
  }

  private OnNewsRecyclerItemClickListener<News> onRecyclerItemClickListener;

  @Override public void setOnItemClickListener(OnNewsRecyclerItemClickListener<News> listener) {
    onRecyclerItemClickListener = listener;
  }

  final static class NewsViewHolder extends BaseViewHolder {
    @BindView(R.id.title_textview) TextView titleTextview;
    @BindView(R.id.count_textview) TextView countTextview;
    @BindView(R.id.date_textview) TextView dateTextView;
    @BindView(R.id.simple_news_imageview) ImageView imageView;

    public NewsViewHolder(View itemView) {
      super(itemView);
    }
  }
}

File: app/src/main/java/io/github/ovso/righttoknow/ui/main/news/NewsAdapter.kt
package io.github.ovso.righttoknow.ui.main.news

import android.text.SpannableString
import android.view.View
import androidx.core.text.HtmlCompat
import io.github.ovso.righttoknow.R
import io.github.ovso.righttoknow.framework.adapter.BaseAdapterView
import io.github.ovso.righttoknow.framework.adapter.BaseRecyclerAdapter
import io.github.ovso.righttoknow.framework.utils.Utility
import io.github.ovso.righttoknow.ui.main.news.model.News
import kotlinx.android.extensions.LayoutContainer
import kotlinx.android.synthetic.main.fragment_news_item.*
import java.util.*

class NewsAdapter : BaseRecyclerAdapter(), NewsAdapterDataModel, BaseAdapterView {
  private val items: MutableList<News> = ArrayList()
  override fun createViewHolder(view: View, viewType: Int): BaseViewHolder {
    return NewsViewHolder(view)
  }

  override fun getLayoutRes(viewType: Int): Int {
    return R.layout.fragment_news_item
  }

  override fun onBindViewHolder(holder: BaseViewHolder, position: Int) {
    if (holder is NewsViewHolder) {
      holder.onBindViewHolder(getItem(position), position)
      holder.itemView.setOnClickListener {
        if (onRecyclerItemClickListener != null) {
          onRecyclerItemClickListener!!.onItemClick(getItem(position))
        }
      }
      holder.simple_news_imageview.setOnClickListener {
        onRecyclerItemClickListener!!.onSimpleNewsItemClick(getItem(position))
      }
    }
  }

  override fun getItemCount(): Int {
    return size
  }

  override fun add(item: News) {
    items.add(item)
  }

  override fun addAll(items: List<News>) {
    this.items.addAll(items)
  }

  override fun remove(position: Int): News {
    return items.removeAt(position)
  }

  override fun getItem(position: Int): News {
    return items[position]
  }

  override fun add(index: Int, item: News) {
    items.add(index, item)
  }

  override val size: Int
    get() = items.size

  override fun clear() {
    items.clear()
  }

  override fun refresh() {
    notifyDataSetChanged()
  }

  private var onRecyclerItemClickListener: OnNewsRecyclerItemClickListener<News>? = null
  override fun setOnItemClickListener(listener: OnNewsRecyclerItemClickListener<News>) {
    onRecyclerItemClickListener = listener
  }

  internal class NewsViewHolder(override val containerView: View?) : BaseViewHolder(containerView!!), LayoutContainer {

    fun onBindViewHolder(_item: News, position: Int) {
      val title = _item.title
      val spannableString = SpannableString(HtmlCompat.fromHtml(title, HtmlCompat.FROM_HTML_MODE_COMPACT))
      title_textview.text = spannableString
      count_textview.text = (position + 1).toString()
      val date = Utility.convertDate(_item.pubDate, "yy-MM-dd")
      date_textview.text = date
    }
  }
}